content/en/blog/2024/inpod-traffic-redirection-ambient/index.md

```diff
@@ -0,0 +1,228 @@
+the host network based traffic redirection mechanism we had been using for Istio ambient traffic capture and redirection
+during the alpha stage was not going to meet our requirements. 
+
+These are the two biggest challenges with the host-based approach:
+- The default CNI’s node-level networking configuration could interfere with the node-level networking configuration
+from istio-cni, whether it is with eBPF or iptables.
+- If users deploy a network policy for the default CNI, the network policy may not be enforced when istio-cni is deployed.
+
+### Addressing the challenges
+
+Applying any traffic
+routing/networking rules in the node-level network namespace invites unresolvable conflicts/incompatibilities with
+CNIs (which might use a wide variety of fundamentally networking topologies on the node) - they *must* configure traffic
+routing/networking rules in the node-level network namespace. We realized any eBPF implementation would have the same
+basic problem, as there is no standardized way to safely chain/extend arbitrary eBPF programs at this time.
+
+In sidecar mode, it is trivial to configure traffic redirection between sidecar and application pod, as both operate within
+the pod's network namespace. This led to a light-bulb moment: why not mimic sidecars, and configure the redirection in
+the application network namespace? 
+
+While this sounds like a "crazy simple" thought, is this even possible, given ztunnel runs
+in the Istio system namespace? After some research, we discovered a Linux process running in one network namespace
+could create and own listening sockets within another network namespace, which is a basic Linux socket capability.
+However, to make this work, and cover all pod lifecycle scenarios we had to make architectural changes to the ztunnel
+as well as the istio-cni agent.
+
+### Traffic redirection in ambient now
+
+After sufficient prototyping and validating that the innovative approach does work for all the Kubernetes platforms we have
+access to, we built the confidence of the work and contributed to upstream to switch to this new traffic redirection
+model - an *in-Pod* traffic redirection mechanism between workload pods and the ztunnel node proxy component that has
+been built from the ground up to be highly compatible with all major cloud providers and CNIs.
+
+The key innovation is to deliver the pod’s network namespace to the ztunnel so that ztunnel can start its redirection
+sockets inside the pod’s network namespace, while still running outside the pod. With this approach, the traffic redirection
+between ztunnel and application pods happens in a way that’s very similar to sidecars and application pods today and is
+strictly invisible to any Kubernetes CNI. Network policy from any Kubernetes CNIs can continue to be enforced,
+regardless of whether the CNI uses eBPF or iptables, there is no conflict between the in-Pod traffic redirection and the CNI.
```

Comment:
We have seen that this works even with OVN-Kubernetes CNI where Network policies that are implemented using OpenFlow rules. In case you want, you can even include "OpenFlow" in the above sentence.

Reply:
I don't feel strongly this should be included. If you do feel strongly, please suggest a change. Thx!